zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ListenerContainer.java

```diff
@@ -0,0 +1,161 @@
+ * right after its consumption, while the latter need to be explicitly removed.
+ * @param <T> the type of listener
+ */
+public class ListenerContainer<T> {
+  private final Map<String, Set<T>> _persistentListener = new ConcurrentHashMap<>();
+  private final Map<String, Set<T>> _onetimeListener = new ConcurrentHashMap<>();
+
+  /**
+   * Add listener to the container with specified key.
+   * @param key the key to register to
+   * @param listener the listener to register
+   * @param persistListener true if the listener is persistent
+   */
+  public void addListener(String key, T listener, boolean persistListener) {
+    addListener(key, listener, persistListener ? _persistentListener : _onetimeListener);
+  }
+
+  /**
+   * Remove listener from the container.
+   * This operation removes both one-time and persistent listener from the specified key.
+   * @param key the key to remove
+   * @param listener the listener to remove
+   */
+  public synchronized void removeListener(String key, T listener) {
+    removeFromListenerMap(key, listener, _persistentListener);
+    removeFromListenerMap(key, listener, _onetimeListener);
+  }
+
+  /**
+   * Remove listener from the container.
+   * @param key the key to remove
+   * @param listener the listener to remove
+   * @param persistListener true if remove the persistent listener, otherwise remove the one-time listener
+   */
+  public synchronized void removeListener(String key, T listener, boolean persistListener) {
+    if (persistListener) {
+      removeFromListenerMap(key, listener, _persistentListener);
+    } else {
+      removeFromListenerMap(key, listener, _onetimeListener);
+    }
+  }
+
+  /**
+   * Remove listeners from the container.
+   * @param key the key to remove
+   * @param listeners the listeners to remove
+   * @param persistListener true if remove the persistent listener, otherwise remove the one-time listener
+   */
+  public synchronized void removeListeners(String key, Collection<T> listeners, boolean persistListener) {
+    if (persistListener) {
+      removeFromListenerMap(key, listeners, _persistentListener);
+    } else {
+      removeFromListenerMap(key, listeners, _onetimeListener);
+    }
+  }
+
+  /**
+   * Consume the listeners registered to the given key. The given consumer is applied to all one-time and persistent
+   * listeners with the given key.
+   * The one-time listeners are removed after this operation, as a result, this operation is NOT idempotent.
+   * @param key the key of the listeners
+   * @param consumer the consuming action for the listeners
+   */
+  public void consumeListeners(String key, Consumer<? super T> consumer) {
+    Set<T> onetimeListeners;
+    Set<T> persistentListeners;
+    synchronized (this) {
+      // remove one-time listeners
+      onetimeListeners = _onetimeListener.remove(key);
+    }
+    if (onetimeListeners == null) {
```

Comment:
I feel that from line 106 to line 114 can be simplified, such that only if it's non-null then wen add it to the result set, and there's no need of instantiating empty set when there's no listener registered.

Reply:
Thanks for the review. 
To my understanding, the reference to `Collections.emptySet()` won't instantiate any new instance, so there is no overhead. I think this is also good for readability and unify the control logic with non-null values. 
Let me know if you have other thoughts on this.